node/network/approval-distribution/src/lib.rs

```diff
@@ -1622,30 +1901,108 @@ impl State {
 						lag = ?self.approval_checking_lag,
 						"Encountered old block pending gossip topology",
 					);
-					return
+					return *required_routing
 				}
 
+				let mut new_required_routing = *required_routing;
+
 				if config.l1_threshold.as_ref().map_or(false, |t| &self.approval_checking_lag >= t)
 				{
 					// Message originator sends to everyone.
-					if local && *required_routing != RequiredRouting::All {
+					if local && new_required_routing != RequiredRouting::All {
 						metrics.on_aggression_l1();
-						*required_routing = RequiredRouting::All;
+						new_required_routing = RequiredRouting::All;
 					}
 				}
 
 				if config.l2_threshold.as_ref().map_or(false, |t| &self.approval_checking_lag >= t)
 				{
 					// Message originator sends to everyone. Everyone else sends to XY.
-					if !local && *required_routing != RequiredRouting::GridXY {
+					if !local && new_required_routing != RequiredRouting::GridXY {
 						metrics.on_aggression_l2();
-						*required_routing = RequiredRouting::GridXY;
+						new_required_routing = RequiredRouting::GridXY;
 					}
 				}
+				new_required_routing
 			},
+			&self.peer_views,
 		)
 		.await;
 	}
+
+	// Filter out invalid candidate index and certificate core bitfields.
+	// For each invalid assignment we also punish the peer.
+	async fn sanitize_v1_assignments(
+		&mut self,
+		peer_id: PeerId,
+		sender: &mut impl overseer::ApprovalDistributionSenderTrait,
+		assignments: Vec<(IndirectAssignmentCert, CandidateIndex)>,
+	) -> Vec<(IndirectAssignmentCertV2, CandidateBitfield)> {
+		let mut sanitized_assignments = Vec::new();
+		for (cert, candidate_index) in assignments.into_iter() {
+			let cert_bitfield_bits = match cert.cert.kind {
+				AssignmentCertKind::RelayVRFDelay { core_index } => core_index.0 as usize + 1,
+				// We don't want to run the VRF yet, but the output is always bounded by `n_cores`.
+				// We assume `candidate_bitfield` length for the core bitfield and we just check against
+				// `MAX_BITFIELD_SIZE` later.
+				AssignmentCertKind::RelayVRFModulo { .. } => candidate_index as usize + 1,
+			};
+
+			// Ensure bitfields length under hard limit.
+			if cert_bitfield_bits > MAX_BITFIELD_SIZE ||
+				cert_bitfield_bits != candidate_index as usize + 1
```

Comment:
```suggestion
				
```

This check is not correct candidate_index is not the same as core_index for RelayVRFDelay, discovered this while running in a loop this test: https://github.com/paritytech/polkadot/pull/7554/files#diff-91b440e49b019790cb20969ee1f5d91731280bbd3de896dc25835ed7474212d2R35

Reply:
yes indeed, good catch!